### PR TITLE
Adds logic of having a last sync timestamp for each TEI - DHIS2-3830

### PR DESCRIPTION
**Pull request by huzvak**
Every TEI now holds an information about when it was last synced to the master server.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/TrackerSynchronization.java [hunk 1]
```diff
@@ -125,15 +126,20 @@ public void syncTrackerProgramData()
                 log.debug( "TEIs that are going to be synced are: " + dtoTeis );
             }
 
-            if ( !sendTrackerSyncRequest( dtoTeis, username, password ) )
+            if ( sendTrackerSyncRequest( dtoTeis, username, password ) )
+            {
+                List<String> teiUIDs = dtoTeis.stream().map( TrackedEntityInstance::getTrackedEntityInstance ).collect( Collectors.toList() );
+                log.info( "The lastSynced flag of these TEIs should be updated to: " + teiUIDs );
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/sync/TrackerSynchronization.java]: remove "to".

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceStore.java [hunk 2]
```diff
@@ -64,4 +65,12 @@
...
      */
     boolean existsIncludingDeleted( String uid );
+
+    /**
+     * Set lastSynced timestamp to provided timestamp for provided TEIs
+     *
+     * @param entityInstanceUIDs UIDs of Tracked entity instances where the lastSynced flag should be updated
+     * @param lastSynced         The date of last successful sync
+     */
+    void updateTrackedEntityInstancesSyncTimestamp( List<String> entityInstanceUIDs, Date lastSynced );
```
Comment (stian-sandvold) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceStore.java]: entityInstanceUIDs -> trackedEntityInstanceUIDs

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceService.java [hunk 3]
```diff
@@ -202,6 +210,14 @@ void validate( TrackedEntityInstanceQueryParams params )
      */
     void updateTrackedEntityInstance( TrackedEntityInstance entityInstance );
 
+    /**
+     * Updates a last sync timestamp on specified TrackedEntityInstances
+     *
+     * @param entityInstanceUIDs UIDs of Tracked entity instances where the lastSynced flag should be updated
+     * @param lastSynced         The date of last successful sync
+     */
+    void updateTrackedEntityInstancesSyncTimestamp( List<String> entityInstanceUIDs, Date lastSynced );
```
Comment (stian-sandvold) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstanceService.java]: entityInstanceUIDs -> trackedEntityInstanceUIDs

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstance.java [hunk 4]
```diff
@@ -79,6 +80,8 @@
 
     private String coordinates;
 
+    private Date lastSynced = new Date( 0 );
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityInstance.java]: Perhaps rename to "lastSynchronized".

A bit more formal and consistent as i think we use this other places.

**Pull request merged**